File: gluecodium/src/test/resources/smoke/equatable/output/cpp/src/smoke/SimpleEquatableStruct.cpp
// -------------------------------------------------------------------------------------------------
//
//
// -------------------------------------------------------------------------------------------------
#include "smoke/NonEquatableClass.h"
#include "smoke/NonEquatableInterface.h"
#include "smoke/SimpleEquatableStruct.h"
#include <utility>
namespace smoke {
SimpleEquatableStruct::SimpleEquatableStruct( )
    : class_field{ }, interface_field{ }, nullable_class_field{ }, nullable_interface_field{ }
{
}
SimpleEquatableStruct::SimpleEquatableStruct( ::std::shared_ptr< ::smoke::NonEquatableClass > class_field, ::std::shared_ptr< ::smoke::NonEquatableInterface > interface_field, ::std::shared_ptr< ::smoke::NonEquatableClass > nullable_class_field, ::std::shared_ptr< ::smoke::NonEquatableInterface > nullable_interface_field )
    : class_field( std::move( class_field ) ), interface_field( std::move( interface_field ) ), nullable_class_field( std::move( nullable_class_field ) ), nullable_interface_field( std::move( nullable_interface_field ) )
{
}
bool SimpleEquatableStruct::operator==( const SimpleEquatableStruct& rhs ) const
{
    return class_field == rhs.class_field &&
        interface_field == rhs.interface_field &&
        ( ( nullable_class_field && rhs.nullable_class_field )
            ? ( &*nullable_class_field == &*rhs.nullable_class_field )
            : ( static_cast< bool >( nullable_class_field ) == static_cast< bool >( rhs.nullable_class_field ) ) ) &&
        ( ( nullable_interface_field && rhs.nullable_interface_field )
            ? ( &*nullable_interface_field == &*rhs.nullable_interface_field )
            : ( static_cast< bool >( nullable_interface_field ) == static_cast< bool >( rhs.nullable_interface_field ) ) );
}
bool SimpleEquatableStruct::operator!=( const SimpleEquatableStruct& rhs ) const
{
    return !( *this == rhs );
}
}
namespace gluecodium {
std::size_t
hash< ::smoke::SimpleEquatableStruct >::operator( )( const ::smoke::SimpleEquatableStruct& t ) const
{
    size_t hash_value = 43;
    hash_value = ( hash_value ^ ::gluecodium::hash< decltype( ::smoke::SimpleEquatableStruct::class_field ) >( )( t.class_field ) ) << 1;
hash_value = ( hash_value ^ ::gluecodium::hash< decltype( ::smoke::SimpleEquatableStruct::interface_field ) >( )( t.interface_field ) ) << 1;
hash_value = ( hash_value ^ ::gluecodium::hash< decltype( ::smoke::SimpleEquatableStruct::nullable_class_field ) >( )( t.nullable_class_field ) ) << 1;
hash_value = ( hash_value ^ ::gluecodium::hash< decltype( ::smoke::SimpleEquatableStruct::nullable_interface_field ) >( )( t.nullable_interface_field ) ) << 1;
    return hash_value;
}
}

File: gluecodium/src/test/resources/smoke/equatable/output/swift/smoke/SimpleEquatableStruct.swift
//
//
import Foundation
public struct SimpleEquatableStruct: Hashable {
    public var classField: NonEquatableClass
    public var interfaceField: NonEquatableInterface
    public var nullableClassField: NonEquatableClass?
    public var nullableInterfaceField: NonEquatableInterface?
    public init(classField: NonEquatableClass, interfaceField: NonEquatableInterface, nullableClassField: NonEquatableClass? = nil, nullableInterfaceField: NonEquatableInterface? = nil) {
        self.classField = classField
        self.interfaceField = interfaceField
        self.nullableClassField = nullableClassField
        self.nullableInterfaceField = nullableInterfaceField
    }
    internal init(cHandle: _baseRef) {
        classField = NonEquatableClass_moveFromCType(smoke_SimpleEquatableStruct_classField_get(cHandle))
        interfaceField = NonEquatableInterface_moveFromCType(smoke_SimpleEquatableStruct_interfaceField_get(cHandle))
        nullableClassField = NonEquatableClass_moveFromCType(smoke_SimpleEquatableStruct_nullableClassField_get(cHandle))
        nullableInterfaceField = NonEquatableInterface_moveFromCType(smoke_SimpleEquatableStruct_nullableInterfaceField_get(cHandle))
    }
    /// :nodoc:
    public static func == (lhs: SimpleEquatableStruct, rhs: SimpleEquatableStruct) -> Bool {
        return
            lhs.classField === rhs.classField &&
            lhs.interfaceField === rhs.interfaceField &&
            lhs.nullableClassField === rhs.nullableClassField &&
            lhs.nullableInterfaceField === rhs.nullableInterfaceField
    }
    /// :nodoc:
    public func hash(into hasher: inout Hasher) {
        hasher.combine(Unmanaged<AnyObject>.passUnretained(classField).toOpaque().hashValue)
        hasher.combine(Unmanaged<AnyObject>.passUnretained(interfaceField).toOpaque().hashValue)
        hasher.combine(nullableClassField != nil ? Unmanaged<AnyObject>.passUnretained(nullableClassField!).toOpaque().hashValue : 0)
        hasher.combine(nullableInterfaceField != nil ? Unmanaged<AnyObject>.passUnretained(nullableInterfaceField!).toOpaque().hashValue : 0)
    }
}
internal func copyFromCType(_ handle: _baseRef) -> SimpleEquatableStruct {
    return SimpleEquatableStruct(cHandle: handle)
}
internal func moveFromCType(_ handle: _baseRef) -> SimpleEquatableStruct {
    defer {
        smoke_SimpleEquatableStruct_release_handle(handle)
    }
    return copyFromCType(handle)
}
internal func copyToCType(_ swiftType: SimpleEquatableStruct) -> RefHolder {
    let c_classField = moveToCType(swiftType.classField)
    let c_interfaceField = moveToCType(swiftType.interfaceField)
    let c_nullableClassField = moveToCType(swiftType.nullableClassField)
    let c_nullableInterfaceField = moveToCType(swiftType.nullableInterfaceField)
    return RefHolder(smoke_SimpleEquatableStruct_create_handle(c_classField.ref, c_interfaceField.ref, c_nullableClassField.ref, c_nullableInterfaceField.ref))
}
internal func moveToCType(_ swiftType: SimpleEquatableStruct) -> RefHolder {
    return RefHolder(ref: copyToCType(swiftType).ref, release: smoke_SimpleEquatableStruct_release_handle)
}
internal func copyFromCType(_ handle: _baseRef) -> SimpleEquatableStruct? {
    guard handle != 0 else {
        return nil
    }
    let unwrappedHandle = smoke_SimpleEquatableStruct_unwrap_optional_handle(handle)
    return SimpleEquatableStruct(cHandle: unwrappedHandle) as SimpleEquatableStruct
}
internal func moveFromCType(_ handle: _baseRef) -> SimpleEquatableStruct? {
    defer {
        smoke_SimpleEquatableStruct_release_optional_handle(handle)
    }
    return copyFromCType(handle)
}
internal func copyToCType(_ swiftType: SimpleEquatableStruct?) -> RefHolder {
    guard let swiftType = swiftType else {
        return RefHolder(0)
    }
    let c_classField = moveToCType(swiftType.classField)
    let c_interfaceField = moveToCType(swiftType.interfaceField)
    let c_nullableClassField = moveToCType(swiftType.nullableClassField)
    let c_nullableInterfaceField = moveToCType(swiftType.nullableInterfaceField)
    return RefHolder(smoke_SimpleEquatableStruct_create_optional_handle(c_classField.ref, c_interfaceField.ref, c_nullableClassField.ref, c_nullableInterfaceField.ref))
}
internal func moveToCType(_ swiftType: SimpleEquatableStruct?) -> RefHolder {
    return RefHolder(ref: copyToCType(swiftType).ref, release: smoke_SimpleEquatableStruct_release_optional_handle)
}
